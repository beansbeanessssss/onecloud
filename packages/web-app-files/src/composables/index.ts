export * from './actions'
export * from './resourcesViewDefaults'
export * from './shares'
