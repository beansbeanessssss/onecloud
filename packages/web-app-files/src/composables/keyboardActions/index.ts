export * from './useKeyboardTableNavigation'
export * from './useKeyboardTableMouseActions'
export * from './useKeyboardTableActions'
export * from './useKeyboardTableSpaceActions'
