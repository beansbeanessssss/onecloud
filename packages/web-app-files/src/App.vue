<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="oc-width-expand" />
  </main>
</template>
<script lang="ts">
import { defineComponent, watch } from '@vue/composition-api'
import { Route } from 'vue-router'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { bus } from 'web-pkg/src/instance'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    watch(useRoute(), (to: Route, from?: Route) => {
      store.dispatch('Files/resetFileSelection')
    })
  },
  data: () => ({
    dragareaEnabled: false
  }),
  created() {
    const dragOver = bus.subscribe('drag-over', this.onDragOver)
    const dragOut = bus.subscribe('drag-out', this.hideDropzone)
    const drop = bus.subscribe('drop', this.hideDropzone)

    this.$on('beforeDestroy', () => {
      bus.unsubscribe('drag-over', dragOver)
      bus.unsubscribe('drag-out', dragOut)
      bus.unsubscribe('drop', drop)
    })
  },

  methods: {
    hideDropzone() {
      this.dragareaEnabled = false
    },
    onDragOver(event) {
      this.dragareaEnabled = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
