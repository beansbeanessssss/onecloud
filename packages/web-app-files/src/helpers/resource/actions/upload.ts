import filesize from 'filesize'
import { Resource } from 'web-client'
import {
  extractExtensionFromFile,
  isShareSpaceResource,
  SpaceResource
} from 'web-client/src/helpers'
import { CreateDirectoryTreeResult, UppyResource } from 'web-runtime/src/composables/upload'
import { UppyService } from 'web-runtime/src/services/uppyService'
import { ConflictDialog, ResolveConflict, resolveFileNameDuplicate, ResolveStrategy } from '..'
import { locationPublicLink } from '../../../router/public'
import { locationSpacesGeneric } from '../../../router/spaces'

export class ResourcesUpload extends ConflictDialog {
  constructor(
    private filesToUpload: File[],
    private currentFiles: Resource[],
    private inputFilesToUppyFiles: (inputFileOptions) => UppyResource[],
    private $uppyService: UppyService,
    private space: SpaceResource,
    private currentFolder: string,
    private currentFolderId: string | number,
    private spaces: SpaceResource[],
    private hasSpaces: boolean,
    private createDirectoryTree: (
      space: SpaceResource,
      currentPath: string,
      files: UppyResource[],
      currentFolderId?: string | number
    ) => Promise<CreateDirectoryTreeResult>,
    createModal: (modal: object) => void,
    hideModal: () => void,
    showMessage: (data: object) => void,
    $gettext: (msg: string) => string,
    $ngettext: (msgid: string, plural: string, n: number) => string,
    $gettextInterpolate: (msg: string, context: object, disableHtmlEscaping?: boolean) => string
  ) {
    super(createModal, hideModal, showMessage, $gettext, $ngettext, $gettextInterpolate)
  }

  async perform() {
    const conflicts = []
    const uppyResources: UppyResource[] = this.inputFilesToUppyFiles(this.filesToUpload)
    const quotaExceeded = this.checkQuotaExceeded(uppyResources)

    if (quotaExceeded) {
      return this.$uppyService.clearInputs()
    }
    for (const file of uppyResources) {
      const relativeFilePath = file.meta.relativePath
      if (relativeFilePath) {
        // Logic for folders, applies to all files inside folder and subfolders
        const rootFolder = relativeFilePath.replace(/^\/+/, '').split('/')[0]
        const exists = this.currentFiles.find((f) => f.name === rootFolder)
        if (exists) {
          if (conflicts.some((conflict) => conflict.name === rootFolder)) {
            continue
          }
          conflicts.push({
            name: rootFolder,
            type: 'folder'
          })
          continue
        }
      }
      // Logic for files
      const exists = this.currentFiles.find(
        (f) => f.name === file.name && !file.meta.relativeFolder
      )
      if (exists) {
        conflicts.push({
          name: file.name,
          type: 'file'
        })
      }
    }
    if (conflicts.length) {
      await this.displayOverwriteDialog(uppyResources, conflicts)
    } else {
      this.handleUppyFileUpload(this.space, this.currentFolder, uppyResources)
    }
  }

  async handleUppyFileUpload(space: SpaceResource, currentFolder: string, files: UppyResource[]) {
    this.$uppyService.publish('uploadStarted')
    const result = await this.createDirectoryTree(space, currentFolder, files, this.currentFolderId)

    let filesToUpload = files
    if (result.failed.length) {
      filesToUpload = files.filter(
        (f) => !result.failed.some((r) => f.meta.relativeFolder.startsWith(r))
      )
    }

    if (filesToUpload.length) {
      this.$uppyService.publish('addedForUpload', filesToUpload)
      this.$uppyService.uploadFiles(filesToUpload)
    } else {
      this.$uppyService.publish('uploadCompleted', { successful: [] })
    }
  }

  async displayOverwriteDialog(files: UppyResource[], conflicts) {
    let count = 0
    const allConflictsCount = conflicts.length
    const resolvedFileConflicts = []
    const resolvedFolderConflicts = []
    let doForAllConflicts = false
    let allConflictsStrategy
    let doForAllConflictsFolders = false
    let allConflictsStrategyFolders

    for (const conflict of conflicts) {
      const isFolder = conflict.type === 'folder'
      const conflictArray = isFolder ? resolvedFolderConflicts : resolvedFileConflicts

      if (doForAllConflicts && !isFolder) {
        conflictArray.push({
          name: conflict.name,
          strategy: allConflictsStrategy
        })
        continue
      }
      if (doForAllConflictsFolders && isFolder) {
        conflictArray.push({
          name: conflict.name,
          strategy: allConflictsStrategyFolders
        })
        continue
      }
      const conflictsLeft = allConflictsCount - count
      const resolvedConflict: ResolveConflict = await this.resolveFileExists(
        { name: conflict.name, isFolder } as Resource,
        conflictsLeft,
        conflictsLeft === 1,
        isFolder
      )
      count++
      if (resolvedConflict.doForAllConflicts) {
        if (isFolder) {
          doForAllConflictsFolders = true
          allConflictsStrategyFolders = resolvedConflict.strategy
        } else {
          doForAllConflicts = true
          allConflictsStrategy = resolvedConflict.strategy
        }
      }

      conflictArray.push({
        name: conflict.name,
        strategy: resolvedConflict.strategy
      })
    }
    const filesToSkip = resolvedFileConflicts
      .filter((e) => e.strategy === ResolveStrategy.SKIP)
      .map((e) => e.name)
    const foldersToSkip = resolvedFolderConflicts
      .filter((e) => e.strategy === ResolveStrategy.SKIP)
      .map((e) => e.name)

    files = files.filter((e) => !filesToSkip.includes(e.name))
    files = files.filter(
      (file) =>
        !foldersToSkip.some((folderName) => file.meta.relativeFolder.split('/')[1] === folderName)
    )

    const filesToKeepBoth = resolvedFileConflicts
      .filter((e) => e.strategy === ResolveStrategy.KEEP_BOTH)
      .map((e) => e.name)
    const foldersToKeepBoth = resolvedFolderConflicts
      .filter((e) => e.strategy === ResolveStrategy.KEEP_BOTH)
      .map((e) => e.name)

    for (const fileName of filesToKeepBoth) {
      const file = files.find((e) => e.name === fileName && !e.meta.relativeFolder)
      const extension = extractExtensionFromFile({ name: fileName } as Resource)
      file.name = resolveFileNameDuplicate(fileName, extension, this.currentFiles)
    }
    for (const folder of foldersToKeepBoth) {
      const filesInFolder = files.filter((e) => e.meta.relativeFolder.split('/')[1] === folder)
      for (const file of filesInFolder) {
        const newFolderName = resolveFileNameDuplicate(folder, '', this.currentFiles)
        file.meta.relativeFolder = file.meta.relativeFolder.replace(
          new RegExp(`/${folder}`),
          `/${newFolderName}`
        )
        file.meta.relativePath = file.meta.relativePath.replace(
          new RegExp(`/${folder}/`),
          `/${newFolderName}/`
        )
        file.meta.tusEndpoint = file.meta.tusEndpoint.replace(
          new RegExp(`/${folder}`),
          `/${newFolderName}`
        )
      }
    }
    if (files.length === 0) {
      return this.$uppyService.clearInputs()
    }
    this.handleUppyFileUpload(this.space, this.currentFolder, files)
  }

  checkQuotaExceeded(uppyResources: UppyResource[]) {
    if (!this.hasSpaces) {
      return false
    }

    let quotaExceeded = false

    const uploadSizeSpaceMapping = uppyResources.reduce((acc, uppyResource) => {
      let targetUploadSpace

      if (uppyResource.meta.routeName === locationPublicLink.name) {
        return acc
      }

      if (uppyResource.meta.routeName === locationSpacesGeneric.name) {
        targetUploadSpace = this.spaces.find((space) => space.id === uppyResource.meta.spaceId)
      }

      if (!targetUploadSpace || isShareSpaceResource(targetUploadSpace)) {
        return acc
      }

      const matchingMappingRecord = acc.find(
        (mappingRecord) => mappingRecord.space.id === targetUploadSpace.id
      )

      if (!matchingMappingRecord) {
        acc.push({
          space: targetUploadSpace,
          uploadSize: uppyResource.data.size
        })
        return acc
      }

      matchingMappingRecord.uploadSize += uppyResource.data.size

      return acc
    }, [])

    uploadSizeSpaceMapping.forEach(({ space, uploadSize }) => {
      if (space.spaceQuota.remaining && space.spaceQuota.remaining < uploadSize) {
        let spaceName = space.name

        if (space.driveType === 'personal') {
          spaceName = this.$gettext('Personal')
        }

        const translated = this.$gettext(
          'There is not enough quota on %{spaceName}, you need additional %{missingSpace} to upload these files'
        )

        this.showMessage({
          title: this.$gettext('Not enough quota'),
          desc: this.$gettextInterpolate(translated, {
            spaceName,
            missingSpace: filesize((space.spaceQuota.remaining - uploadSize) * -1)
          }),
          status: 'danger'
        })

        quotaExceeded = true
      }
    })

    return quotaExceeded
  }
}
