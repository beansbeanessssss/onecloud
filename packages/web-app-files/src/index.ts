import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import { defineWebApplication, quickActions } from '@ownclouders/web-pkg'
import store from './store'
import { extensions } from './extensions'
import fileSideBars from './fileSideBars'
import { buildRoutes } from '@ownclouders/web-pkg'
import { AppNavigationItem } from '@ownclouders/web-pkg'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import {
  SpaceResource,
  isPersonalSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  color: 'var(--oc-color-swatch-primary-muted)',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
export const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      isActive: () => {
        return (
          !context.$store.getters['runtime/spaces/currentSpace'] ||
          context.$store.getters['runtime/spaces/currentSpace']?.isOwner(
            context.$store.getters.user
          )
        )
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find(
          (drive) => isPersonalSpaceResource(drive) && drive.isOwner(context.$store.getters.user)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      isActive: () => {
        const space = context.$store.getters['runtime/spaces/currentSpace'] as SpaceResource
        // last check is when fullShareOwnerPaths is enabled
        return !space || isShareSpaceResource(space) || !space?.isOwner(context.$store.getters.user)
      },
      activeFor: [
        { path: `/${appInfo.id}/spaces/share` },
        { path: `/${appInfo.id}/spaces/personal` }
      ],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0' && capabilities.files?.undelete
      }
    }
  ]
}

export default defineWebApplication({
  setup(args) {
    return {
      appInfo,
      store,
      routes: buildRoutes({
        App,
        Favorites,
        FilesDrop,
        SearchResults,
        Shares: {
          SharedViaLink,
          SharedWithMe,
          SharedWithOthers
        },
        Spaces: {
          DriveResolver: SpaceDriveResolver,
          Projects: SpaceProjects
        },
        Trash: {
          Overview: TrashOverview
        }
      }),
      navItems,
      quickActions,
      translations,
      extensions: extensions(args)
    }
  }
})
