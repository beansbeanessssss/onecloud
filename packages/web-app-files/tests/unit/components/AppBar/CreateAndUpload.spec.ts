import CreateAndUpload from 'web-app-files/src/components/AppBar/CreateAndUpload.vue'
import { mock, mockDeep } from 'jest-mock-extended'
import { Resource, SpaceResource } from 'web-client/src/helpers'
import { UppyResource } from 'web-runtime/src/composables/upload'
import { Graph } from 'web-client'
import { Drive } from 'web-client/src/generated'
import { eventBus, useRequest } from 'web-pkg'
import {
  createStore,
  defaultPlugins,
  shallowMount,
  defaultStoreMockOptions,
  defaultComponentMocks
} from 'web-test-helpers'

jest.mock('web-pkg/src/composables/authContext')

const elSelector = {
  component: '#create-and-upload-actions',
  newFileButton: '#new-file-menu-btn',
  uploadBtn: '#upload-menu-btn',
  resourceUpload: 'resource-upload-stub',
  newFolderBtn: '#new-folder-btn',
  clipboardBtns: '#clipboard-btns',
  pasteFilesBtn: '.paste-files-btn',
  clearClipboardBtn: '.clear-clipboard-btn'
}

const fileHandlerMocks = [
  {
    ext: 'txt',
    action: {
      app: 'text-editor',
      newTab: false,
      extension: 'txt'
    },
    menuTitle: () => 'Plain text file'
  },
  {
    ext: 'md',
    action: {
      app: 'text-editor',
      newTab: false,
      extension: 'md'
    },
    menuTitle: () => 'Mark-down file'
  },
  {
    ext: 'drawio',
    action: {
      app: 'draw-io',
      newTab: true,
      routeName: 'draw-io-edit',
      extension: 'drawio'
    },
    menuTitle: () => 'Draw.io document'
  }
]

describe('CreateAndUpload component', () => {
  describe('action buttons', () => {
    it('should show and be enabled if file creation is possible', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.findComponent<any>(elSelector.uploadBtn).props().disabled).toBeFalsy()
      expect(wrapper.findComponent<any>(elSelector.newFolderBtn).props().disabled).toBeFalsy()
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('should be disabled if file creation is not possible', () => {
      const currentFolder = mockDeep<Resource>({ canUpload: () => false })
      const { wrapper } = getWrapper({ currentFolder })
      expect(wrapper.findComponent<any>(elSelector.uploadBtn).props().disabled).toBeTruthy()
      expect(wrapper.findComponent<any>(elSelector.newFolderBtn).props().disabled).toBeTruthy()
    })
    it('should not be visible if file creation is not possible on a public page', () => {
      const currentFolder = mockDeep<Resource>({ canUpload: () => false })
      const { wrapper } = getWrapper({ currentFolder, currentRouteName: 'files-public-link' })
      expect(wrapper.find(elSelector.component).exists()).toBeFalsy()
    })
  })
  describe('file handlers', () => {
    it('should always show for uploading files and folders', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.findAll(elSelector.resourceUpload).length).toBe(2)
    })
    it('should show additional handlers', () => {
      const { wrapper } = getWrapper({ newFileHandlers: fileHandlerMocks })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('clipboard buttons', () => {
    it('should show if clipboard is empty', () => {
      const { wrapper } = getWrapper()
      expect(
        wrapper.findComponent<any>(`${elSelector.clipboardBtns} oc-button-stub`).exists()
      ).toBeFalsy()
    })
    it('should show if clipboard is not empty', () => {
      const { wrapper } = getWrapper({ clipboardResources: [mockDeep<Resource>()] })
      expect(wrapper.findAll(`${elSelector.clipboardBtns} .oc-button`).length).toBe(2)
    })
    it('call the "paste files"-action', async () => {
      const { wrapper, storeOptions } = getWrapper({ clipboardResources: [mockDeep<Resource>()] })
      await wrapper.find(elSelector.pasteFilesBtn).trigger('click')
      expect(storeOptions.modules.Files.actions.pasteSelectedFiles).toHaveBeenCalled()
    })
    it('call "clear clipboard"-action', async () => {
      const { wrapper, storeOptions } = getWrapper({ clipboardResources: [mockDeep<Resource>()] })
      await wrapper.find(elSelector.clearClipboardBtn).trigger('click')
      expect(storeOptions.modules.Files.actions.clearClipboardFiles).toHaveBeenCalled()
    })
  })
  describe('button triggers', () => {
    it('should create a modal if "New folder" button is clicked', async () => {
      const { wrapper } = getWrapper()
      const createModalSpy = jest.spyOn(wrapper.vm, 'createModal')
      await wrapper.find(elSelector.newFolderBtn).trigger('click')
      expect(createModalSpy).toHaveBeenCalled()
    })
    it.each(fileHandlerMocks)(
      'should create a modal if "New file" button is clicked',
      async (fileHandler) => {
        const { wrapper } = getWrapper({ newFileHandlers: fileHandlerMocks })
        const createModalSpy = jest.spyOn(wrapper.vm, 'createModal')
        const showCreateResourceModal = jest.spyOn(wrapper.vm, 'showCreateResourceModal')
        await wrapper.find(`.new-file-btn-${fileHandler.ext}`).trigger('click')
        expect(createModalSpy).toHaveBeenCalled()
        expect(showCreateResourceModal).toHaveBeenCalledWith(
          false,
          fileHandler.ext,
          fileHandler.action
        )
      }
    )
  })
  describe('method "onUploadComplete"', () => {
    it.each([
      { driveType: 'personal', updated: 1 },
      { driveType: 'project', updated: 1 },
      { driveType: 'share', updated: 0 },
      { driveType: 'public', updated: 0 }
    ])('updates the space quota for supported drive types', async (data) => {
      const { driveType, updated } = data
      const { wrapper, mocks, storeOptions } = getWrapper()
      const file = mockDeep<UppyResource>({ meta: { driveType } })
      const graphMock = mockDeep<Graph>()
      graphMock.drives.getDrive.mockResolvedValue(mockDeep<Drive>() as any)
      mocks.$clientService.graphAuthenticated.mockImplementation(() => graphMock)
      await wrapper.vm.onUploadComplete({ successful: [file] })
      expect(
        storeOptions.modules.runtime.modules.spaces.mutations.UPDATE_SPACE_FIELD
      ).toHaveBeenCalledTimes(updated)
    })
    it('reloads the file list if files were uploaded to the current path', async () => {
      const eventSpy = jest.spyOn(eventBus, 'publish')
      const itemId = 'itemId'
      const space = mock<SpaceResource>({ id: '1' })
      const { wrapper, mocks } = getWrapper({ itemId, space })
      const file = mockDeep<UppyResource>({
        meta: { driveType: 'project', spaceId: space.id, currentFolderId: itemId }
      })
      const graphMock = mockDeep<Graph>()
      graphMock.drives.getDrive.mockResolvedValue(mockDeep<Drive>() as any)
      mocks.$clientService.graphAuthenticated.mockImplementation(() => graphMock)
      await wrapper.vm.onUploadComplete({ successful: [file] })
      expect(eventSpy).toHaveBeenCalled()
    })
  })
  describe('methods "addNewFolder" & "addNewFile"', () => {
    it.each(['addNewFolder', 'addNewFile'])(
      'updates the resource after the folder as been created successfully',
      async (method) => {
        const createMethod = method === 'addNewFolder' ? 'createFolder' : 'putFileContents'
        const { wrapper, mocks, storeOptions } = getWrapper({ item: '/' })
        mocks.$clientService.webdav[createMethod].mockResolvedValue(mockDeep<Resource>())
        await wrapper.vm[method]('New resource')
        expect(storeOptions.modules.Files.mutations.UPSERT_RESOURCE).toHaveBeenCalled()
      }
    )
    it.each(['addNewFolder', 'addNewFile'])(
      'shows a message when an error occurres',
      async (method) => {
        const createMethod = method === 'addNewFolder' ? 'createFolder' : 'putFileContents'
        jest.spyOn(console, 'error').mockImplementation(() => undefined)
        const { wrapper, mocks, storeOptions } = getWrapper({ item: '/' })
        mocks.$clientService.webdav[createMethod].mockRejectedValue(new Error())
        await wrapper.vm[method]('New resource')
        expect(storeOptions.actions.showMessage).toHaveBeenCalled()
      }
    )
    it('opens the file editor after a file has been created and a supported editor is available', async () => {
      const { wrapper, mocks } = getWrapper({
        newFileAction: true,
        item: '/',
        newFileHandlers: fileHandlerMocks
      })
      const openEditorSpy = jest.spyOn(wrapper.vm, '$_fileActions_openEditor').mockImplementation()
      mocks.$clientService.webdav.putFileContents.mockResolvedValue(mockDeep<Resource>())
      await wrapper.vm.addNewFile('New resource.txt')
      expect(openEditorSpy).toHaveBeenCalled()
    })
  })
  describe('methods "checkNewFolderName" & "checkNewFileName"', () => {
    it.each([
      { name: '', valid: false },
      { name: '/name', valid: false },
      { name: '.', valid: false },
      { name: '..', valid: false },
      { name: 'name ', valid: false },
      { name: 'name', valid: true }
    ])('verifies the resource name', (data) => {
      const { name, valid } = data
      const { wrapper } = getWrapper()
      const folderResult = wrapper.vm.checkNewFolderName(name)
      expect(folderResult === null).toBe(valid)
      const fileResult = wrapper.vm.checkNewFileName(name)
      expect(fileResult === null).toBe(valid)
    })
    it('shows error when the resource name already exists', () => {
      const existingFile = mockDeep<Resource>({ name: 'someFile.txt' })
      const { wrapper } = getWrapper({ files: [existingFile] })
      const folderResult = wrapper.vm.checkNewFolderName(existingFile.name)
      expect(folderResult).not.toBeNull()
      const fileResult = wrapper.vm.checkNewFileName(existingFile.name)
      expect(fileResult).not.toBeNull()
    })
  })
  describe('method "addAppProviderFile"', () => {
    it('triggers the default file action', async () => {
      const { wrapper, mocks } = getWrapper({ item: '/' })
      const defaultActionSpy = jest
        .spyOn(wrapper.vm, '$_fileActions_triggerDefaultAction')
        .mockImplementation(() => undefined)
      mocks.$clientService.webdav.getFileInfo.mockResolvedValue(mockDeep<Resource>())
      await wrapper.vm.addAppProviderFile('someFile.txt')
      expect(defaultActionSpy).toHaveBeenCalled()
    })
    it('shows a message when an error occurred', async () => {
      jest.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper, mocks, storeOptions } = getWrapper({ item: '/' })
      mocks.$clientService.webdav.getFileInfo.mockRejectedValue(new Error())
      await wrapper.vm.addAppProviderFile('someFile.txt')
      expect(storeOptions.actions.showMessage).toHaveBeenCalled()
    })
  })
})

function getWrapper({
  newFileHandlers = [],
  clipboardResources = [],
  files = [],
  currentFolder = mockDeep<Resource>({ canUpload: () => true }),
  currentRouteName = 'files-spaces-generic',
  space = mock<SpaceResource>(),
  item = undefined,
  itemId = undefined,
  newFileAction = false
} = {}) {
  jest.mocked(useRequest).mockImplementation(() => ({
    makeRequest: jest.fn().mockResolvedValue({ status: 200 })
  }))
  const storeOptions = {
    ...defaultStoreMockOptions,
    getters: {
      ...defaultStoreMockOptions.getters,
      newFileHandlers: () => newFileHandlers,
      user: () => ({ id: '1' }),
      capabilities: () => ({
        spaces: { enabled: true },
        files: { app_providers: [{ new_url: '/' }] }
      })
    }
  }
  storeOptions.getters.apps.mockImplementation(() => ({
    fileEditors: []
  }))
  storeOptions.modules.Files.getters.currentFolder.mockImplementation(() => currentFolder)
  storeOptions.modules.Files.getters.clipboardResources.mockImplementation(() => clipboardResources)
  storeOptions.modules.Files.getters.files.mockImplementation(() => files)
  const store = createStore(storeOptions)
  const mocks = { ...defaultComponentMocks({ currentRoute: { name: currentRouteName } }) }
  return {
    storeOptions,
    mocks,
    wrapper: shallowMount(CreateAndUpload as any, {
      data: () => ({ newFileAction }),
      props: { space: space as any, item, itemId },
      global: {
        stubs: { OcButton: false },
        renderStubDefaultSlot: true,
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
