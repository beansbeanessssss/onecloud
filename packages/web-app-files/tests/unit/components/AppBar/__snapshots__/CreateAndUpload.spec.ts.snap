// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAndUpload component action buttons should show and be enabled if file creation is possible 1`] = `
<div class="create-and-upload-actions oc-flex-inline oc-width-1-1"><span><oc-button-stub type="button" size="medium" arialabel="Create a new folder" submit="button" variation="primary" appearance="filled" justifycontent="center" gapsize="medium" id="new-folder-btn"><oc-icon-stub name="resource-type-folder" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>New Folder</span></oc-button-stub></span> <span><oc-button-stub type="button" size="medium" arialabel="Upload files or folders" submit="button" variation="passive" appearance="outline" justifycontent="center" gapsize="medium" id="upload-menu-btn"><oc-icon-stub name="upload" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>Upload</span></oc-button-stub></span>
  <oc-drop-stub dropid="upload-menu-drop" toggle="#upload-menu-btn" position="bottom-start" mode="click" closeonclick="true" paddingsize="small" class="oc-width-auto">
    <oc-list-stub id="upload-list">
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1"></resource-upload-stub>
      </li>
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1" isfolder="true"></resource-upload-stub>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
  <div id="clipboard-btns" class="oc-button-group">
    <!---->
    <!---->
  </div>
</div>
`;

exports[`CreateAndUpload component file handlers should show additional handlers 1`] = `
<div class="create-and-upload-actions oc-flex-inline oc-width-1-1"><span><oc-button-stub type="button" size="medium" arialabel="Create new files or folders" submit="button" variation="primary" appearance="filled" justifycontent="center" gapsize="medium" id="new-file-menu-btn"><oc-icon-stub name="add" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>New</span></oc-button-stub></span>
  <oc-drop-stub dropid="new-file-menu-drop" toggle="#new-file-menu-btn" position="bottom-start" mode="click" closeonclick="true" paddingsize="small" class="oc-width-auto">
    <oc-list-stub id="create-list">
      <li class="create-list-folder oc-menu-item-hover">
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" id="new-folder-btn">
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub>
          <translate-stub tag="span">Folder</translate-stub>
        </oc-button-stub>
      </li>
      <li class="create-list-file oc-menu-item-hover">
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="new-file-btn-txt">
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub> <span>Plain text file</span>
        </oc-button-stub>
      </li>
      <li class="create-list-file oc-menu-item-hover">
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="new-file-btn-md">
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub> <span>Mark-down file</span>
        </oc-button-stub>
      </li>
      <li class="create-list-file oc-menu-item-hover">
        <oc-button-stub type="button" size="medium" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" class="new-file-btn-drawio">
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub> <span>Draw.io document</span>
        </oc-button-stub>
      </li>
    </oc-list-stub>
  </oc-drop-stub> <span><oc-button-stub type="button" size="medium" arialabel="Upload files or folders" submit="button" variation="passive" appearance="outline" justifycontent="center" gapsize="medium" id="upload-menu-btn"><oc-icon-stub name="upload" filltype="line" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub> <span>Upload</span></oc-button-stub></span>
  <oc-drop-stub dropid="upload-menu-drop" toggle="#upload-menu-btn" position="bottom-start" mode="click" closeonclick="true" paddingsize="small" class="oc-width-auto">
    <oc-list-stub id="upload-list">
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1"></resource-upload-stub>
      </li>
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnlabel="" btnclass="oc-width-1-1" isfolder="true"></resource-upload-stub>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
  <div id="clipboard-btns" class="oc-button-group">
    <!---->
    <!---->
  </div>
</div>
`;
